### DISCUSSION TITLE
take copilot suggestion word by word

### DISCUSSION THREAD

**Discussion by nhrnhr0** [community/community#7637] [Copilot Conversations]
Hello, I love copilot! it saves me so much time!
one thing I'm not sure how to do in copilot.
sometimes I want only some of the autocomplete but I can only click TAB and get all the autocomplete, or start typing it myself.
for example:
![copilot](https://user-images.githubusercontent.com/25978588/141476822-63d722ef-066a-474c-9658-9672f032bddd.PNG)
I'm very happy with the result until the verbose_name, but I can only TAB and take all, or start typing `models.TextField(...`
is there a way, for example, to press the control and move with the arrow keys char by char or word by word? So I don't need to type myself the correct part or delete half the suggestion?

**Comment by itsdhung** [comment]
I have a similar suggestion related to completions that span multiple lines. Oftentimes the first line is correct, but then subsequent lines won't be what I want. Since I have no way to accept just the first line, it renders the entire completion useless.

Good workarounds (for my particular coding style) would be:
1. Add a user configuration option to turn off multiline completions. Outside of the issue described above, I also find long multiline completions hard to read. I also personally don't have a problem completing the first line → typing a few chars → completing the next line → typing a few chars → completing the next line → etc...
2. Something similar to what @nhrnhr0 is suggesting where there's an interaction that allows the user to complete just the first line of a multiline completion.

**Comment by MulleDK19** [comment]
Hit Tab, then Ctrl+Backspace till you get to verbose_name, update it, and let it continue the suggestions.

**Reply by omartheman** [reply]
This is precisely what OP said they did not want to have to do - go back and delete half of the suggestion.

**Reply by ekharchenko-avito** [reply]
"half" is a good option. often it is a multiline suggestion where only a few first words are correct.

**Comment by evanthebouncy** [comment]
surprised this isn't a feature already, I guess we'll wait until they work on it more, it's literally 1 hack away, and I'd say a few undergrads with a hackathon and the cost of 5 slices of pizzas would just fix this

**Comment by yangjuncode** [comment]
I love this cause the whold suggestion is too much code i don't need sometimes

**Comment by Erik-Schlegel** [comment]
Would love to see this feature implemented as well. Found this because I had the exact same suggestion and googled to see if it was already a thing.

**Reply by wybert** [reply]
That's a great idea. I have used copilot for a while, and really love it but sometimes I just want to accept part of the suggestions like one line or one function.

**Comment by wybert** [comment]
I think it will be also AWESOME if copilot can allow users to click the end of the acceptance from the whole suggestion.

**Reply by docwilco** [reply]
While I like the idea of one click to mark the end, I'm not a fan of moving my hand to a mouse while coding.

I think the idea of ctrl/alt and arrow keys is pretty good. Especially if up/down would navigate a whole line, sorta like selecting with shift. Left/right would do a whole word though, for speed. I think you need to be able to go back, since overshooting is common.

**Reply by wybert** [reply]
That sounds even better!

**Reply by cmp-nct** [reply]
By now I'd say anything, really just anything is better than the excessive code dumps.
I'm really scared of damaging vital source code by missing out one line when deleting the completion dumps.

**Comment by pidgeon777** [comment]
I would love autocompleting `copilot` suggestions word by word, eventually by adding a special shortcut for that.

Also see:

https://github.com/community/community/discussions/16299

**Comment by cmp-nct** [comment]
Almost a year later.
They won't fix that because it destroys their marketing.
When you accept 12 lines of code because you wanted 0.3 lines, then delete 12.7 lines of code their marketing metric says: Copilot just wrote 12 lines of your code, that's 60% efficiency.

If they would allow people to complete word by word they would actually be able to work efficiently with it but their marketing would crumble.

Given that it's a year now, anyone hacked the plugins to do that ?
I looked into the jetbrains one and it can be decompiled. I also located the area "actions" and "inlay" are the keywords but I have no experience with java or plugins. So it's quite annoying to go that route

**Comment by omartheman** [comment]
I would think a faulty marketing metric isn’t really a good business model

On Sun, Oct 16, 2022 at 8:25 PM John ***@***.***> wrote:

> Almost a year later.
> They won't fix that because it destroys their marketing.
> When you accept 12 lines of code because you wanted 0.3 lines, then delete
> 12.7 lines of code their marketing metric says: Copilot just wrote 12 lines
> of your code, that's 60% efficiency.
>
> If they would allow people to complete word by word they would actually be
> able to work efficiently with it but their marketing would crumble.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/community/discussions/7637#discussioncomment-3891794>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMBIQ5O26RMSPZVVNT3ULQDWDSMGZANCNFSM5H44HVGA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
*Omar Shishani*
*Eclipse Software*
(707) 342 - 8648 <+17073428648>
https://omarshishani.com

<https://omarshishani.com/>

**Reply by cmp-nct** [reply]
It pisses me off every day I use it, it would be such a time saver but I spend more time fixing the excessive completions than writing code now.
Though if you look at their marketing, they focus on exactly that fallacy. 
We'll likely need to hack the plugin to solve it.

**Comment by omartheman** [comment]
Well eventually a competitor will come in and make a better product if they
don’t fix it

On Sun, Oct 16, 2022 at 8:28 PM John ***@***.***> wrote:

> It pisses me off every day I use it, it would be such a time saver but I
> spend more time fixing the excessive completions than writing code now.
> Though if you look at their marketing, they focus on exactly that fallacy.
> We'll likely need to hack the plugin to solve it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/community/discussions/7637#discussioncomment-3891808>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMBIQ5KJV34WNSQR477U2PTWDSMTJANCNFSM5H44HVGA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
*Omar Shishani*
*Eclipse Software*
(707) 342 - 8648 <+17073428648>
https://omarshishani.com

<https://omarshishani.com/>

**Reply by cmp-nct** [reply]
Doubtless, though it requires quite an infrastructure to provide the quick answers for such a network.
There are already competitors and in a year or two there will be equal ones I guess.

Until then I think we need to solve it on our own.
I am not into Plugin development, the cleanest solution would be to make a plugin that immediately deletes an accepted completion and offers it again, this time with proper control on inserting.
If no one can do the plugin I'll maybe find an hour to hack something together in C# or Autohotkeys

**Comment by cmp-nct** [comment]
I wrote a plugin that takes over the completion and offers it word-by-word
It splits up the completion into words, you can adapt the delimiter (by default white space, comma and new line)
https://github.com/cmp-nct/Stewardess
Finally I can use copilot with precision.

There are likely still bugs, the code handling the carret/line/document offsets isn't too robust.
However, it works for me.

**Reply by TomBener** [reply]
Any plan to support VS Code?

**Reply by cmp-nct** [reply]
I didn't even know Kotlin some days ago and never made a plugin (and I now know I dislike both), creating that thing was torture.
However, it's quite readable. Maybe someone improves it further and/or ports it.

I also added a feature that shows what copilot is doing by analyzing it's internal status and looking for changes.
Also quite useful, it shows if copilot is actually working after requesting a completion or if it's frozen.

**Reply by drwilco** [reply]
I looked into the VS Code side of things, and unfortunately the VS Code CoPilot extension does not export any data/functions. Since VS Code itself does not allow extensions to look at Decorations, it is impossible to write something properly at this point.

I am tempted to try and shim the CoPilot extension to proxy the setDecorations call to get the data, but unsure how GitHub will like that

**Comment by krishnapraveen7** [comment]
Yo, if anyone attempted to make a plugin using copilot to solve this issue , that would be hilarious.

**Reply by cmp-nct** [reply]
I used copilot a lot to create the plugin, I did not even know the programming language involved (Kotlin) and never did a plugin before.
It certainly helped, though enduring the flaws of Copilot in that timespan while I wrote a plugin to solve those flaws nearly drove me insane :)

**Comment by docwilco** [comment]
Great news, everyone!

Github Copilot for VSCode just uses the suggestion API for VSCode. And as of version 1.75 of VSCode (January 2023) there are `editor.action.inlineSuggest.acceptNextWord` and `editor.action.inlineSuggest.undo` which are bound to ctrl-right and ctrl-left respectively.

Bad news is that there's no next line and previous line options yet, but they might be in the works.

Ironically, I discovered this completely by accident... while working on a VSCode extension to do this sort of thing. 

Issue for next/prev line has already been filed here: https://github.com/microsoft/vscode/issues/173962

**Reply by docwilco** [reply]
And here is my VSCode extension: https://marketplace.visualstudio.com/items?itemName=drwilco.change-stepper

The difference between it and the native suggestion stepping is that you have to accept the whole suggestion first for mine, but then you can go line by line _and_ word by word, both forwards and backwards!

**Comment by bksubramanyarao** [comment]
It's been more than one year, and still for all JetBrains products, I am missing this very important and critical feature, Partial World Accept. When VS Code already has it, why not JetBrains? I don't want to install any other third-party plugin. When I'm paying for software, it looks like the team at Copilot are sleeping.
